# COVID19_heritability
Heritability estimates for COVID-19 HGI phenotypes. 

SNP heritability was estimated via the GenomicSEM implementation of LDSC using only EUR summary statistics, a EUR LD reference panel, the sum of the effective sample size (the sample size for an equivalently powered GWAS within a balanced sample—i.e., 50% cases and 50% controls) across the contributing cohorts, a sample prevalence of 0.5, and across a range of population-based prevalence (1%–90%). Using the summation of the effective sample size across cohorts, rather than total sample size or effective sample size calculated using total sample prevalence, accounts for differences in sample prevalence across contributing cohorts in a meta-analysis. Estimating heritability without accounting for differences in ascertainment can result in a downward bias by as much as 30%. SNP heritability for all three COVID-19 related phenotypes was significant on the observed-scale for all the three phenotypes (1.2–8.2%, P < 0.0001). Liability-scale heritabilities were estimated to range between 0.0066–0.013, 0.032–0.061, and 0.045–0.086 for reported infection, hospitalization, and critical illness, respectively.

For me details on using GenomicSEM for estimating heritability using the sum of effective sample sizes, see their [wiki](https://github.com/GenomicSEM/GenomicSEM/wiki/2.1-Calculating-Sum-of-Effective-Sample-Size-and-Preparing-GWAS-Summary-Statistics) and [corresponding paper](https://www.sciencedirect.com/science/article/pii/S0006322322013166?via%3Dihub).  
